bpf/custom/bpf_custom.c

```diff
@@ -0,0 +1,29 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+/* Copyright (C) 2020 Authors of Cilium */
+
+#include <bpf/ctx/skb.h>
+#include <bpf/api.h>
+
+#include "lib/common.h"
+
+/* Include custom header file (.h) containing the implementation for
+ * custom_prog(), for example:
+ *
+ *     #include "bytecount.h"
+ */
+#include "bytecount.h"
+
+__section("custom")
+int custom_hook(struct __ctx_buff *ctx)
+{
+	int ret = ctx_load_meta(ctx, CB_CUSTOM_CALLS);
+	int identity = get_identity(ctx);
+
+	/* Call user-defined function from custom header file. */
+	custom_prog(ctx, identity);
```

Comment:
Do we want to use a more specific name for this hook? We may need to consider this name part of the API because changing it will require (small) fixes for users of this new feature. If we later add new custom hooks in the datapath, we will need to add new functions similar to this one (and can't call them all `custom_prog` :slightly_smiling_face: ).

Reply:
`custom_prog` has the advantage of being quite explicit for the review :) But agreed, we can probably find something else before we merge. A simple alternative would be to rename `bpf/custom/bpf_custom.c` into `bpf/custom_metrics/bpf_custom_metrics.c`, and `custom_prog()` function into something like `collect_metrics()`? Or do you have something else in mind?

Reply:
I think we should clarify the position of the hook in the datapath somehow because its function (e.g., collect metrics) really depends on what users will do with it. So maybe something like `post_container_policy`?

Reply:
But then if we had other hooks elsewhere in the future, we could theoretically reuse the same `bpf_custom.c` as a landing point, but then the name would not apply anymore? (Thinking of it, it also applies to something with `custom_metrics` in its name). Hmm :thinking:

Reply:
> But then if we had other hooks elsewhere in the future, we could theoretically reuse the same bpf_custom.c as a landing point

That was actually my initial concern. I don't think we'll want to reuse that function for other (new) hooks because otherwise existing custom programs will start receiving a lot more events (i.e., packets) than they expect. If the program is counting packets for example, it may all of a sudden start counting packets twice or start counting packets it doesn't want to count.

Reply:
I'm not sure I'm following you. You can reuse the `bpf_custom.c` base, but to produce a different programme that would be attached to another hook. If the user-defined part is different, they should be no risk of having packets counted multiple times?